### PR TITLE
Add support for Socket Logging Handler

### LINKED ISSUE #23127: Logging into socket
### Description

Hello, 
I would like to add option to use plain UDP/TCP/SSL_TCP socket logging which is supported out of the box by JBossLogManager. In our environment we dont have currently allowed to use GELF nor Syslog. We were able to log from our app to logstash but those 2 input methods are not currently allowed.

We would like to add basic formatters like plain and JSON.
In Json: we would like to add custom fields to log message which can be setup 'statically' during launch time. Exactly as GELF with additional fields. We also need option to override keys as some fields are differently named in StructuredFormatter.

### Configuration suggestion

quarkus.log.handler.socket.enabled=true
quarkus.log.handler.socket.host=localhost
quarkus.log.handler.socket.port=12201
quarkus.log.handler.socket.protocol=UDP
quarkus.log.handler.socket.formatter=json
quarkus.log.handler.socket.keyoverrides=LEVEL=severity
quarkus.log.handler.socket.additional-field.serverName.value=myServerName

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by JozefDropco**
Resolves https://github.com/quarkusio/quarkus/issues/23127

### CODE REVIEW

**Comment by geoand** [comment]
Hi, 

Thanks for this!

I think the changes are more complicated than necessary. Personally I would prefer something like: https://github.com/geoand/quarkus/tree/%2323127

**Comment by JozefDropco** [comment]
Hi @geoand 
thanks I will update it. give me some time and I will update it

**Comment by geoand** [comment]
No rush

File: core/runtime/src/main/java/io/quarkus/runtime/logging/SocketConfig.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+     * TCP} or {@link Protocol#SSL_TCP SSL TCP} protocol
+     */
+    @ConfigItem
+    boolean blockOnReconnect;
+
+    /**
+     * The log message formatter. json or pattern is supported
+     */
+    @ConfigItem(defaultValue = "pattern")
+    String formatter;
```
Comment (geoand) [inline core/runtime/src/main/java/io/quarkus/runtime/logging/SocketConfig.java:48]: Why is this added vs what SysLogConfig?

Comment (JozefDropco) [reply core/runtime/src/main/java/io/quarkus/runtime/logging/SocketConfig.java:48]: This is added as the output where you are sending messages can require formatted output differently than the Pattern. In our case we have logstash which has input tcp and json_lines. We need custom attributes/fields to add to the json file. If you have some other suggestion how to achieve it - I am happy to modify it.

Comment (geoand) [reply core/runtime/src/main/java/io/quarkus/runtime/logging/SocketConfig.java:48]: @gsmet WDYT about this? Does this use case warrant doing something different than what our SysLog handling does?

Comment (JozefDropco) [reply core/runtime/src/main/java/io/quarkus/runtime/logging/SocketConfig.java:48]: Initially I thought I will use logging-json extension however this extension is limited to console logging. Maybe it would be more beneficial to extend that support instead and keep socket logging clean

Comment (geoand) [reply core/runtime/src/main/java/io/quarkus/runtime/logging/SocketConfig.java:48]: Yeah, I think that's a better approach

Comment (JozefDropco) [reply core/runtime/src/main/java/io/quarkus/runtime/logging/SocketConfig.java]: HI @geoand, @gsmet ,
I have reworked the code completely. Now socket logger in core has just pattern and logging-json was extended to support all handlers not just console as it was. Now file, console, syslog, socket and also named handlers can log as json. Previous implementation of consoleFormatter was opiniated and if JSON was added as extension basic console but also all named handlers were forced to log as JSON. I kept the original behavior but personally I am not fan of it. With "json=false"  you can set console to use PatternFormatter and some named console handlers to use JSON. But you cannot use console as JSON and some named console handler to use Pattern. Its just because implementation should be backward compatible.
If users add logging-json extension 
- all handlers will start log to JSON (even file,syslog,socket) . Not sure if its ok?
- Additional fields can be defined as quarkus.log.socket.json.additional-field.myownfield.value=testingvalue 
- Keyoverrides was exposed too

I changed integration test of logging-json as previous one was not doing anything.

Comment (geoand) [reply core/runtime/src/main/java/io/quarkus/runtime/logging/SocketConfig.java]: Thanks a lot @JozefDropco.

I'll have a look soon

**Comment by geoand** [comment]
@JozefDropco we can't remove build items as this PR does as that could break other extensions

**Comment by JozefDropco** [comment]
Hi @geoand ,
ok what do you suggest than? keep original extension as it was and introduce new build item + new extension which will be using the new code?

**Comment by geoand** [comment]
I'll have to find some spare time to look into it and let you know.

Thanks for your work on this and for your patience

**Comment by geoand** [comment]
Lets say this PR didn't exist, what is your end goal here? 
I am trying to understand what the reason is for all the changes, as they are certainly not all necessary for implementing https://github.com/quarkusio/quarkus/issues/23127 (as my https://github.com/quarkusio/quarkus/compare/main...geoand:%2323127?expand=1).

Can you please state your end goal(s) clearly?

Thanks

**Comment by JozefDropco** [comment]
HI @geoand,
our goal is to be able to log into logstash under TCP with SSL enabled. We need to structure the messages as JSON with some custom fields. Index in logstash is not managed by us. 

I am able to log to kibana with logback without any issues but this is true for the application. But not for the server. Quarkus is using JUL (jboss implementation with MDC support). but we wouldnt like to write our application dependeing on server directly. We would like to use SLF4J with logback. When I was trying I was unable to force quarkus to use our impelmentation instead jboss-logmanager in DEV mode. When I build it as uber-jar it was working correctly. Is there a specific reason why it wasnt working?

**Comment by geoand** [comment]
So in order to make things absolutely clear, what you need is support for the Socket logging handler along with JSON support, correct?

If so, then I think we'll need to figure out some changes to allow JSON support in other handlers as well

**Comment by JozefDropco** [comment]
Hi, honestly the ConsoleFormatter build item is quite hardcoded and I don't
see any reason why somebody would need to log as json in console unless
this output is not sent via filebeat somewhere else.  Nevertheless I was
able to hack it. In application I use slf4j with logback (with tcp
appender). I use Command to start server from main method where I reset
handlers of JUL and bind it to slf4j. Now all logs are flowing through
logback to logstash. Anyways I am opened to implement even more
straightforward solution if you are opened to accept it

Kind regards, Jozef

Dňa st 2. 2. 2022, 16:17 Georgios Andrianakis ***@***.***>
napísal(a):

**Comment by geoand** [comment]
I agree with you that currently the json support is very hard coded and we should probably make it possible to have json support across all handlers.

The reason to log to json in the console is because in Kubernetes environments applications usually just log to the console and have the platform ship the logs wherever is necessary.

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/LogConsoleFormatBuildItem.java [hunk 2]
```diff
@@ -1,36 +0,0 @@
...
-
-import io.quarkus.builder.item.MultiBuildItem;
-import io.quarkus.runtime.RuntimeValue;
-
-/**
- * The log console format build item. Producing this item will cause the logging subsystem to disregard its
- * console logging formatting configuration and use the formatter provided instead. If multiple formatters
- * are enabled at run time, a warning message is printed and only one is used.
- */
-public final class LogConsoleFormatBuildItem extends MultiBuildItem {
```
Comment (gastaldi) [inline core/deployment/src/main/java/io/quarkus/deployment/builditem/LogConsoleFormatBuildItem.java:16]: 👎🏻  Removing this build item will break other extensions

**Review by gastaldi — CHANGES_REQUESTED** [review]

File: core/runtime/src/main/java/io/quarkus/runtime/logging/SocketConfig.java (hunk 1 above)
Comment (gsmet) [reply core/runtime/src/main/java/io/quarkus/runtime/logging/SocketConfig.java]: I somehow missed this PR but I wonder if the work I did in https://github.com/quarkusio/quarkus/pull/25505 isn't somehow related.
I probably need to have another look at it.

**Comment by gsmet** [comment]
I'm going to close this. We made several changes in how we can support JSON in other logging transports (which also led to this PR conflicting a lot) so it's probably a good idea to revisit the issue with a fresh mind and make sure we agree on the approach and on what's needed.

Thanks for the effort anyway. Feel free to ping me to have a discussion about how to pursue this. I will be happy to provide the infrastructure so that the implementation is easier on your side.
